icefoganalytics/travel-authorization
# Part 5: Wire Travel Desk Request Tab

Relates to:

- Part of solving https://github.com/icefoganalytics/travel-authorization/issues/167
- [TravelDeskWireFrame.pdf](https://github.com/icefoganalytics/travel-authorization/files/14016343/TravelDeskWireFrame.pdf)
- [UserStories-TravelDesk.docx](https://github.com/icefoganalytics/travel-authorization/files/14016344/UserStories-TravelDesk.docx)
- [TravelDeskDataModel (1).xlsx](https://github.com/icefoganalytics/travel-authorization/files/14016345/TravelDeskDataModel.1.xlsx)
- https://github.com/ytgov/travel-authorization/tree/main/web/src/modules/travelDesk
- https://github.com/icefoganalytics/travel-authorization/pull/182
- https://github.com/icefoganalytics/travel-authorization/pull/183
- https://github.com/icefoganalytics/travel-authorization/pull/185
- https://github.com/icefoganalytics/travel-authorization/pull/187

# Context

Using the existing Travel Desk forms (https://github.com/ytgov/travel-authorization/tree/main/web/src/modules/travelDesk) rewire the travel desk request to exist in the travel desk tab of the travel auth.

Travel Desk forms should be created anytime a user is traveling by air.

![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/74adc941-50cd-4219-a723-062822c4a94a)

# Implementation

- Standardize travel desk other transportation table and switch to Sequelize model backed access.
- Implement travel desk travel request stateful "submit" action. Redirects to (non-implemented) read-only travel desk travel request page.
- Build composable travel desk other transportation edit table.
- Implement event pipeline to refresh tab state when user triggers an action that changes tab state. Identified some issues with the tab pattern I'm currently using. Will need a separate PR to fix.

# Screenshots

Other Transportation Request edit table
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/143543e6-21e9-48c4-a306-eede5aa6ef3f)

Add Transportation modal with validation triggered by submit.
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/8206fb73-be27-4483-9828-a888212d8054)

Edit Transportation modal
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/be74c054-39d4-44f1-8e73-8c0aa1af9489)

Other Transportation table with some content
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/9e9774a9-f07a-43b7-aa2a-2c9f72a0b11b)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Go to the "My Travel Requests" page via the top drop down nav.
5. Create a travel authorization.
6. In the details section, select at least one travel method with "aircraft" type.
7. Generate the travel estimate.
8. Submit the travel authorization to yourself as the supervisor.
9. Go to Manger View page accessible via the drop down nav.
10. Click on the new request from the "Pending Approvals" panel in the top left.
11. Scroll to the bottom and approve the request.
12. Go back to the "My Travel Requests" page, and find your new travel request in the table.
13. Click on the Submit Travel Desk Request button to go to the Request tab.
14. Scroll down to the Other Transportation Request section and click the "Add Transportation" button to open the add transportation dialog.
15. Click "Add" before entering any content to trigger validation, and check that appropriate fields are required.
16. Fill in the appropriate fields and then click "Add" to save the other transportation request.
17. Check that record gets added to the other transportation table.
18. Edit the request, via the pencil icon on the table row, check that everything works as requested and the request saves.
19. Check that the table refreshes after save, and gets the newly updated request.
20. Check that you can delete requests via the red "x" in the table row.
21. Scroll to the button of the page and click the submit button.
22. Check that the page redirects to the my-travel-requests -> request page (page has not been implemented).
23. Check that the tab has updated to reflect that you can only read the travel desk travel request.
24. Check that using the "back" button does not take you to the edit version of the page. (router.replace)
